pkg/pool-weighted/test/YieldProtocolFees.test.ts

```diff
@@ -0,0 +1,188 @@
+import { expect } from 'chai';
+import { BigNumber, Contract } from 'ethers';
+import { deploy } from '@balancer-labs/v2-helpers/src/contract';
+import { fp } from '@balancer-labs/v2-helpers/src/numbers';
+import Vault from '@balancer-labs/v2-helpers/src/models/vault/Vault';
+import TokenList from '@balancer-labs/v2-helpers/src/models/tokens/TokenList';
+import { ZERO_ADDRESS } from '@balancer-labs/v2-helpers/src/constants';
+import { random, range } from 'lodash';
+import { toNormalizedWeights } from '@balancer-labs/balancer-js';
+import { calculateInvariant } from '@balancer-labs/v2-helpers/src/models/pools/weighted/math';
+import { actionId } from '@balancer-labs/v2-helpers/src/models/misc/actions';
+import { ProtocolFee } from '@balancer-labs/v2-helpers/src/models/vault/types';
+
+const randomFloat = (min: number, max: number) => random(min, max, true);
+
+describe('YieldProtocolFees', () => {
+  let vault: Vault;
+  let pool: Contract;
+  let rateProviders: Contract[];
+
+  const NAME = 'Balancer Pool Token';
+  const SYMBOL = 'BPT';
+  const POOL_SWAP_FEE_PERCENTAGE = fp(0.01);
+  const PROTOCOL_YIELD_FEE_PERCENTAGE = fp(0.5);
+
+  before('deploy lib', async () => {
+    vault = await Vault.create();
+
+    if (!vault.admin) throw new Error('Vault has no admin');
+    const protocolFeesProvider = vault.protocolFeesProvider;
+    const action = await actionId(protocolFeesProvider, 'setFeeTypePercentage');
+    await vault.grantPermissionsGlobally([action], vault.admin);
+    await protocolFeesProvider
+      .connect(vault.admin)
+      .setFeeTypePercentage(ProtocolFee.YIELD, PROTOCOL_YIELD_FEE_PERCENTAGE);
+  });
+
+  async function deployPool(numTokens: number) {
+    const tokens = await TokenList.create(numTokens, { sorted: true });
+    rateProviders = await tokens.asyncMap(async () => await deploy('v2-pool-utils/MockRateProvider'));
+    const rateProviderAddresses = rateProviders.map((provider) => provider.address);
+
+    pool = await deploy('MockYieldProtocolFees', {
+      args: [
+        vault.address,
+        vault.protocolFeesProvider.address,
+        NAME,
+        SYMBOL,
+        tokens.addresses,
+        rateProviderAddresses,
+        tokens.map(() => ZERO_ADDRESS),
+        POOL_SWAP_FEE_PERCENTAGE,
+        0,
+        0,
+        ZERO_ADDRESS,
+      ],
+    });
+  }
+
+  for (let numTokens = 2; numTokens <= 8; numTokens++) {
+    describe(`for a ${numTokens} token pool`, () => {
+      sharedBeforeEach('deploy pool', async () => {
+        await deployPool(numTokens);
+      });
+
+      describe('constructor', () => {
+        it('sets the rate providers', async () => {
+          const rateProviderAddresses = rateProviders.map((provider) => provider.address);
+          const providers = await pool.getRateProviders();
+
+          expect(providers).to.deep.eq(rateProviderAddresses);
+        });
+      });
```

Comment:
Let's have at least one test for a Pool that pays no fees

Reply:
Added in [34d9384](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1640/commits/34d93843fff277a89b477499f0da4b41f9517526)